sources/nodes/kube.go

```diff
@@ -53,12 +52,41 @@ func (self *kubeNodes) recordGoodNodes(nodes []string) {
 	self.goodNodes = nodes
 }
 
-func parseSelectorOrDie(s string) labels.Selector {
-	selector, err := labels.Parse(s)
-	if err != nil {
-		panic(err)
+func (self *kubeNodes) getNodeInfoAndHostname(node api.Node) (Info, string, error) {
+	nodeInfo := Info{}
+	hostname := ""
+	var nodeErr error
+	for _, addr := range node.Status.Addresses {
+		switch addr.Type {
+		case api.NodeExternalIP:
+			nodeInfo.PublicIP = addr.Address
+		case api.NodeInternalIP:
+			nodeInfo.InternalIP = addr.Address
+		case api.NodeHostName:
+			hostname = addr.Address
+		}
+	}
+	if hostname == "" {
+		hostname = node.Name
+	}
+	if nodeInfo.InternalIP == "" {
+		if hostname == nodeInfo.PublicIP {
+			// If the only identifier we have for the node is a public IP, then use it;
+			// don't force a DNS lookup
+			nodeInfo.InternalIP = nodeInfo.PublicIP
```

Comment:
nit: Can you add a V(4) log here that states that we are using the public IP as-is?